Add Montgomery arithmetic

isislovecruft opened the pull request:
This PR adds arithmetic for projective Montgomery points (as requested in #47), required for the [x25519-dalek](https://github.com/isislovecruft/x25519-dalek) crate.

Please review, then I'll merge the changes to subtle, and then we can remove the temporary cargo git branch dependency on subtle. <3

hdevalence commented on Cargo.toml:
```diff
@@ -35,6 +35,8 @@ version = "0.6"
 version = "^0.2"
 default-features = false
+git = "https://github.com/isislovecruft/subtle"
+branch = "feature/conditional-swap_r1"
```
Assuming this is going to be swapped out when you tag a new subtle release

isislovecruft replied:
I just tagged it like 10 minutes ago, so this is reverted in 49f87816

hdevalence commented on src/constants_32bit.rs:
```diff
@@ -75,6 +75,9 @@ pub const HALF: FieldElement32 = FieldElement32([
 pub const A: FieldElement32       = FieldElement32([
     486662, 0, 0, 0, 0, 0, 0, 0, 0, 0, ]);
 
+/// `APLUS2_OVER_FOUR` is (A+2)/4. (This is used internally within Montgomery laddering.)
```
nitpick: "within the Montgomery ladder"?

isislovecruft replied:
What's weird about saying that?

isislovecruft replied:
Oh, I see, I thought my comment was saying "within the Montgomery ladder" and you were saying that was weird, but instead I'm *not* saying that and it *is* in fact weird.

hdevalence commented on src/decaf.rs:
```diff
@@ -194,6 +194,11 @@ impl<'de> Deserialize<'de> for DecafPoint {
 pub struct DecafPoint(pub ExtendedPoint);
 
 impl DecafPoint {
+    /// Convert this `DecafPoint` to its underlying `ExtendedPoint`.
```
I don't think this conversion should exist, since the underlying point is an implementation detail.

isislovecruft replied:
ack, yeah, removed in d39cb275

hdevalence commented on src/decaf.rs:
```diff
@@ -752,7 +757,7 @@ mod test {
         let compressed_id = CompressedDecaf::identity();
         let id = compressed_id.decompress().unwrap();
-        assert_eq!(id.0.compress_edwards(), CompressedEdwardsY::identity());
+        assert_eq!(id.to_edwards().compress(), CompressedEdwardsY::identity());
```
I think this should just be `0.compress()` since it's OK for the tests to use the internal structure.

isislovecruft replied:
yep, also in d39cb275

hdevalence commented on src/edwards.rs:
```diff
@@ -123,7 +123,6 @@ impl CompressedEdwardsY {
     }
 
     /// Copy this `CompressedEdwardsY` to an array of bytes.
-    /// XXX is this useful?
```
Did we figure out whether it makes sense to have both `as_bytes` and `to_bytes`?

isislovecruft replied:
I think it's nice syntactic sugar, and I'm using both in other crates? Also it is nice because it lets users know it's okay to do something, because anytime you're doing a `.0` somewhere you should feel like you're touching something you shouldn't.

hdevalence commented on src/edwards.rs:
```diff
@@ -447,32 +446,67 @@ impl ProjectivePoint {
+    /// Here, again, to differentiate from points in the twisted Edwards model, we
+    /// call the point `(x,y)` in affine coordinates `(u,v)` and similarly in projective
+    /// space we use `(U:V:W)`.  However, since (as per Montgomery's original work) the
+    /// v-coordinate is superfluous to the definition of the group law, we merely
```
nit: rather than "superfluous", should be something like "scalar multiplication can be computed using only the u-coordinate" since the Montgomery line doesn't inherit the group law from the curve

hdevalence replied:
oops, I see that you already did this below, just here the same change is missing

isislovecruft replied:
Oh, github is showing you an old diff? This was fixed in c08591a7

hdevalence commented on src/field.rs:
```diff
@@ -197,11 +197,15 @@ impl FieldElement {
-    /// XXX should we add a debug_assert that self is nonzero?
+    //
+    // XXX do we want the debug assertion to check for zero? it breaks behaviour
+    // such as that such as in curve25519_dalek::montgomery::test::identity_to_monty.
```
not relevant for this PR, but I guess we should probably specify in all of our code (inv, invsqrt, sqrt_ratio) how we handle zeros

isislovecruft replied:
yeah we should make an issue to note that

hdevalence commented on src/montgomery.rs:
```diff
@@ -65,7 +91,7 @@ impl CompressedMontgomeryU {
     //
     // XXX any other exceptional points for the birational map?
-    pub fn decompress(&self) -> Option<ExtendedPoint> {
+    pub fn decompress_edwards(&self) -> Option<ExtendedPoint> {
```
I don't remember where we'd actually want to use this, instead of just lifting a MontgomeryPoint to an ExtendedPoint... Should we keep it around?

isislovecruft replied:
I kept it around because I wasn't yet fully convinced that the conversion equations we worked out on the board were correct and non-exceptional. I have a git stash where I was working on it, but since the only benefit of that would be speed, I didn't see a reason to remove this until the less kludgy thing is finished properly. Do you feel like Montgomery P² ↦ Edwards P³ conversion would be a blocker for 1.0.0?

hdevalence commented on src/montgomery.rs:
```diff
@@ -84,6 +110,23 @@ impl CompressedMontgomeryU {
+    /// A projective `MontgomeryPoint` corresponding to this compressed point.
+    pub fn decompress(&self) -> MontgomeryPoint {
+        MontgomeryPoint{
+            // XXX is it a problem here if we're not using a canonical encoding? —isis
```
I think the only room for encoding malleability here is from the field elements, and I'm not sure what the problem would be with decoding `2^255 - 18` as `1`... if people need a canonical encoding I guess I think they should use Ristretto?

isislovecruft replied:
Agreed. Do you think I should remove the comment?

isislovecruft replied:
Removed the comment in d39e47ff11

hdevalence commented on src/montgomery.rs:
```diff
@@ -150,34 +193,256 @@ impl CompressedMontgomeryU {
+///
+/// `1` if the points are equal, and `0` otherwise.
+impl Equal for MontgomeryPoint {
+    fn ct_eq(&self, that: &MontgomeryPoint) -> u8 {
```
We can save an inversion here since `(U1:W1) = (U2:W2)` iff `U1*W2 == U2*W1` (since `U1/W1 == U2/W2`)

isislovecruft replied:
Oh, good call. It's now down to ~300ns (from ~27000ns) in commit 5e6e6c3fa8

hdevalence commented on src/montgomery.rs:
```diff
@@ -150,34 +193,256 @@ impl CompressedMontgomeryU {
+///
+/// # Note
+///
+/// All points, except for `(X:W) = (0:0)`, are valid, since the projective
```
nit: I think this should just say that all projective points are valid, and `(0:0)` is the only point of `F_p^2` which does not represent a projective point.

isislovecruft replied:
Okay I reworded it a bit in 7b378ada6

hdevalence commented on src/montgomery.rs:
```diff
@@ -150,34 +193,256 @@ impl CompressedMontgomeryU {
+    /// (an invalid projective point in the Montgomery model).
+    ///
+    /// The doubling case is degenerate, in that using this method to accomplish
+    /// point doubling is less efficient than using `differential_double()`.
```
this can't be used for doubling at all, since then the difference would be the identity...

isislovecruft replied:
ca5b58c2b75

hdevalence commented on src/montgomery.rs:
Does this handle the case `bits[255] = 1` correctly? Can we add a test with a scalar with the highest bit set?

isislovecruft replied:
Added a test in 9da24d8af but I had to mark it with

```rust
#[should_panic(expected = "assertion failed: self[31] <= 127")]
```
since `Scalar::radix_16` has a debug assertion in there to catch it.

isislovecruft commented:
Okay, I think I've addressed all the review issues. Merging for 0.13.0.

Whooooooo, now I can go show off my adorable x25519-dalek documentation which has kittens!!!

hdevalence commented:
LGTM
